File: Swift/Tests/ParserTests/SomeParsersTests.swift
import Testing
import Foundation

@testable import Parser

@Test func testAnyChar(){
	#expect(anyChar.parse("abc").value! == ("a", "bc"))
	#expect(anyChar.parse("").value == nil)
}

@Test func testSatisfy(){
	#expect(satisfy({ $0 == "a" }).parse("abc").value! == ("a", "bc"))
	#expect(satisfy({ $0 == "z" }).parse("abc").value == nil)
	#expect(satisfy({ $0 == "a" }).parse("").value == nil)
}

@Test func testChar(){
	#expect(char("a").parse("abc").value! == ("a", "bc"))
	#expect(char("z").parse("abc").value == nil)
	#expect(char("a").parse("").value == nil)
}

@Test func testEmptyString(){
	#expect(empty_string.parse("abc").value! == ("", "abc"))
}

@Test func testOptional(){
	#expect(optional_c(char("1")).parse("1234").value! == ("1", "234"))
	#expect(optional_c(char("1")).parse("abc").value! == ("", "abc"))
}

@Test func testSpaces(){
	#expect(spaces.parse("abc").value! == ("", "abc"))
	#expect(spaces.parse("   abc").value! == ("   ", "abc"))
}

@Test func testSymbol(){
	#expect(symbol("+").parse(" + abc").value! == ("+", "abc"))
	#expect(symbol("+").parse("abc").value == nil)
}

@Test func testAlnum(){
	#expect(alnum.parse("123abc").value! == ("1", "23abc"))
	#expect(alnum.parse("abc").value! == ("a", "bc"))
	#expect(alnum.parse("_123abc").value! == ("_", "123abc"))
	#expect(alnum.parse("!@#").value == nil)
}

@Test func testName(){
	let psin = name("sin")

	#expect(psin.parse(" sin ").value! == ("sin", ""))
	#expect(psin.parse(" sin (1.)").value! == ("sin", "(1.)"))
	#expect(psin.parse("sinus").value == nil)
}

@Test func testSign(){
	#expect(sign.parse("abc").value! == ("", "abc"))
	#expect(sign.parse("+abc").value! == ("+", "abc"))
	#expect(sign.parse("-abc").value! == ("-", "abc"))

	#expect(usign.parse("abc").value! == ("", "abc"))
	#expect(usign.parse(" + abc").value! == ("+", "abc"))
	#expect(usign.parse(" - abc").value! == ("-", "abc"))
}

@Test func testDigits(){
	#expect(digits.parse("123abc").value! == ("123", "abc"))
	#expect(digits.parse("123  abc").value! == ("123", "  abc"))
	#expect(digits.parse("abc").value! == ("", "abc"))
}

@Test func testDouble(){
	#expect(double.parse("1 abc").value! == (1.0, "abc"))
	#expect(double.parse("1. abc").value! == (1.0, "abc"))
	#expect(double.parse("1.23 abc").value! == (1.23, "abc"))
	#expect(double.parse("-1.23 abc").value == nil) // (-1.23, "abc"))
	#expect(double.parse(".23 abc").value! == (0.23, "abc"))
	#expect(double.parse(" + 1.23 abc").value == nil)
	#expect(double.parse("1.23e10abc").value! == (1.23e10, "abc"))
	#expect(double.parse("1.23e-10abc").value! == (1.23e-10, "abc"))
	#expect(double.parse("abc").value == nil)
}

@Test func testBetween(){
	let expr = between(symbol("("), symbol(")"), { () in double })

	#expect(expr.parse(" ( 123 ) abc").value! == (123.0, "abc"))
	#expect(expr.parse(" ( 123 abc").value == nil)
	#expect(expr.parse(" 123 ) abc").value == nil)
}

@Test func testChainlr1(){
	let add = symbol("+").skip({ Parser.pure({ (x: Double, y: Double) in x + y }) })
	let sub = symbol("-").skip({ Parser.pure({ (x: Double, y: Double) in x - y }) })
	let pow = symbol("^").skip({ Parser.pure({ (x: Double, y: Double) in exp(y * log(x)) }) })

	let pexpr = chainl1(double, add.orElse { () in sub }, false)

	#expect(pexpr.parse("7abc").value! == (7.0, "abc"))
	#expect(pexpr.parse(" 7 - 1 - 2 abc").value! == (4.0, "abc"))
	#expect(pexpr.parse(" 7 - 1 + 2 - 3 abc").value! == (5.0, "abc"))
	#expect(pexpr.parse("abc").value == nil)

	#expect(chainr1(double, pow).parse("3 ^ 2 ^ 3 abc").map({ (round($0.0), $0.1) }).value! == (6561.0, "abc"))
}

File: Rust/src/tests/some_parsers_tests.rs
#[cfg(test)]
mod tests {
    use crate::some_parsers::*;
    use crate::parser::*;

    #[test]
    fn test_any_char() {
        assert_eq!(any_char().parse("abc"), Some(('a', "bc".to_string())));
        assert_eq!(any_char().parse(""), None);
    }

    #[test]
    fn test_satisfy() {
        assert_eq!(satisfy(|c| c == 'a').parse("abc"), Some(('a', "bc".to_string())));
        assert_eq!(satisfy(|c| c == 'z').parse("abc"), None);
        assert_eq!(satisfy(|c| c == 'a').parse(""), None);
    }

    #[test]
    fn test_char() {
        assert_eq!(char_('a').parse("abc"), Some(('a', "bc".to_string())));
        assert_eq!(char_('z').parse("abc"), None);
        assert_eq!(char_('a').parse(""), None);
    }

    #[test]
    fn test_empty_string() {
        assert_eq!(empty_string().parse("abc"), Some(("".to_string(), "abc".to_string())));
    }

    #[test]
    fn test_spaces() {
        assert_eq!(spaces().parse("abc"), Some(("".to_string(), "abc".to_string())));
        assert_eq!(spaces().parse("   abc"), Some(("   ".to_string(), "abc".to_string())));
    }

    #[test]
    fn test_symbol() {
        assert_eq!(symbol('+').parse(" + abc"), Some(('+', "abc".to_string())));
        assert_eq!(symbol('+').parse("abc"), None);
    }

    #[test]
    fn test_alnum() {
        let a = alnum();
        assert_eq!(a.parse("123abc"), Some(('1', "23abc".to_string())));
        assert_eq!(a.parse("_123abc"), Some(('_', "123abc".to_string())));
        assert_eq!(a.parse("!@#"), None)
    }

    #[test]
    fn test_name() {
        let psin = name("sin");

        assert_eq!(psin.parse(" sin "), Some(("sin".to_string(), "".to_string())));
        assert_eq!(psin.parse(" sin (1.)"), Some(("sin".to_string(), "(1.)".to_string())));
        assert_eq!(psin.parse("sinus"), None);
    }

    #[test]
    fn test_optional() {
        assert_eq!(optional_c(char_('1')).parse("1234"), Some(("1".to_string(), "234".to_string())));
        assert_eq!(optional_c(char_('1')).parse("abc"), Some(("".to_string(), "abc".to_string())));
    }

    #[test]
    fn test_sign() {
        let s = sign();
        assert_eq!(s.parse("abc"), Some(("".to_string(), "abc".to_string())));
        assert_eq!(s.parse("+abc"), Some(("+".to_string(), "abc".to_string())));
        assert_eq!(s.parse("-abc"), Some(("-".to_string(), "abc".to_string())));

        let u = usign();
        assert_eq!(u.parse("abc"), Some(("".to_string(), "abc".to_string())));
        assert_eq!(u.parse(" + abc"), Some(("+".to_string(), "abc".to_string())));
        assert_eq!(u.parse(" - abc"), Some(("-".to_string(), "abc".to_string())));
    }

    #[test]
    fn test_digits() {
        let d = digits();
        assert_eq!(d.parse("123abc"), Some(("123".to_string(), "abc".to_string())));
        assert_eq!(d.parse("123  abc"), Some(("123".to_string(), "  abc".to_string())));
        assert_eq!(d.parse("abc"), Some(("".to_string(), "abc".to_string())));
    }

    #[test]
    fn test_double() {
        let d = double();
        assert_eq!(d.parse("1 abc"), Some((1.0, "abc".to_string())));
        assert_eq!(d.parse("1. abc"), Some((1.0, "abc".to_string())));
        assert_eq!(d.parse("1.23 abc"), Some((1.23, "abc".to_string())));
        assert_eq!(d.parse("-1.23 abc"), None);
        assert_eq!(d.parse(".23 abc"), Some((0.23, "abc".to_string())));
        assert_eq!(d.parse(" + 1.23 abc"), None);
        assert_eq!(d.parse("1.23e10abc"), Some((1.23e10, "abc".to_string())));
        assert_eq!(d.parse("1.23e-10abc"), Some((1.23e-10, "abc".to_string())));
        assert_eq!(d.parse("abc"), None);
    }

    #[test]
    fn test_between() {
        let expr = between(symbol('('), symbol(')'), double);

        assert_eq!(expr.parse(" ( 123 ) abc"), Some((123.0, "abc".to_string())));
        assert_eq!(expr.parse(" ( 123 abc"), None);
        assert_eq!(expr.parse(" 123 ) abc"), None);
    }

    #[test]
    fn test_chainlr1() {
        use std::rc::Rc;

        let fadd: Rc<dyn Fn(f64, f64) -> f64> = Rc::new(|x, y| x + y);
        let fsub: Rc<dyn Fn(f64, f64) -> f64> = Rc::new(|x, y| x - y);
        let fpow: Rc<dyn Fn(f64, f64) -> f64> = Rc::new(|x, y| (y * x.ln()).exp());
        let add = symbol('+').skip(|| Parser::pure(fadd.clone()));
        let sub = symbol('-').skip(|| Parser::pure(fsub.clone()));
        let pow = symbol('^').skip(|| Parser::pure(fpow.clone()));

        let pexpr = chainl1(double(), add.or_else_p(sub), false);

        assert_eq!(pexpr.parse("7abc"), Some((7.0, "abc".to_string())));
        assert_eq!(pexpr.parse(" 7 - 1 - 2 abc"), Some((4.0, "abc".to_string())));
        assert_eq!(pexpr.parse(" 7 - 1 + 2 - 3 abc"), Some((5.0, "abc".to_string())));
        assert_eq!(pexpr.parse("abc"), None);

        assert_eq!(chainr1(double(), pow.clone()).parse("3 ^ 2 ^ 3 abc").map(|pair| (pair.0.round(), pair.1)),
            Some((6561.0, "abc".to_string())));
    }
}